### PR TITLE
Improve logging for errors in masking during feature extraction

### PR DESCRIPTION
**Pull request by Ambrosiussen**
The Masking logging is quite unhelpful currently.
There is no indication if the mask arguments used in the Feature Extractor are correct, or if a mask for any given image is missing.
This PR helps visualizing this.

### CODE REVIEW

File: src/colmap/controllers/feature_extraction.cc [hunk 1]
```diff
@@ -279,6 +279,11 @@ class FeatureWriterThread : public Thread {
             image_data.camera.has_prior_focal_length ? " (Prior)" : "");
         LOG(INFO) << StringPrintf("  Features:        %d",
                                   image_data.keypoints.size());
+        if (image_data.mask.Data()) {
+          LOG(INFO) << "  Mask Used:       Yes";
+        } else {
+          LOG(INFO) << "  Mask Used:       No";
+        }
```
Comment (ahojnnes) [inline src/colmap/controllers/feature_extraction.cc:286]: I would guess that 99% of use cases involve no masks. I suggest, let's not clutter the logs in this case.
```suggestion
        if (image_data.mask.Data()) {
          LOG(INFO) << "  Mask:            Yes";
        }
```

Comment (Ambrosiussen) [reply src/colmap/controllers/feature_extraction.cc]: That's a fair assumption to make!

File: src/colmap/controllers/image_reader.cc [hunk 2]
```diff
@@ -146,6 +146,9 @@ ImageReader::Status ImageReader::Next(Camera* camera,
   if (mask && !options_.mask_path.empty()) {
     const std::string mask_path =
         JoinPaths(options_.mask_path, image->Name() + ".png");
+    if (!ExistsFile(mask_path)) {
+      LOG(ERROR) << StringPrintf("Mask Path %s does not exist!", mask_path.c_str());
+    }
     if (ExistsFile(mask_path) && !mask->Read(mask_path, false)) {
       return Status::MASK_ERROR;
```
Comment (ahojnnes) [inline src/colmap/controllers/image_reader.cc:153]: MASK_ERROR already triggers a log message? Is it really necessary to distinguish between file not exists and file is invalid here?

Comment (ahojnnes) [reply src/colmap/controllers/image_reader.cc:153]: See src/colmap/controllers/feature_extraction.cc:265

Comment (Ambrosiussen) [reply src/colmap/controllers/image_reader.cc:153]: Yes, Read() doesn't tell you what the issue is. It simply says it is bad. Now you have a clear error in the log stating what is wrong in the case of a missing file.

Comment (ahojnnes) [reply src/colmap/controllers/image_reader.cc:153]: OK, fair enough. Could you apply a consistent logic in src/colmap/controllers/feature_extraction.cc:335 then?

File: src/colmap/controllers/image_reader.cc [hunk 3]
```diff
@@ -146,6 +146,9 @@ ImageReader::Status ImageReader::Next(Camera* camera,
   if (mask && !options_.mask_path.empty()) {
     const std::string mask_path =
         JoinPaths(options_.mask_path, image->Name() + ".png");
+    if (!ExistsFile(mask_path)) {
+      LOG(ERROR) << StringPrintf("Mask Path %s does not exist!", mask_path.c_str());
+    }
     if (ExistsFile(mask_path) && !mask->Read(mask_path, false)) {
```
Comment (ahojnnes) [inline src/colmap/controllers/image_reader.cc:152]:
```suggestion
    if (!ExistsFile(mask_path)) {
      LOG(ERROR) << "Mask at " << mask_path << " does not exist.";
      return Status::MASK_ERROR;
    }
    if (!mask->Read(mask_path, false)) {
      LOG(ERROR) << "Failed to read invalid mask file at: " << mask_path;
```

Comment (ahojnnes) [reply src/colmap/controllers/image_reader.cc:152]: If we are already making it more verbose, let's be extra precise :-)

**Review by ahojnnes — APPROVED** [review]
Thanks for your contribution. LGTM.

File: src/colmap/controllers/image_reader.cc (hunk 2 above)
Comment (Ambrosiussen) [reply src/colmap/controllers/image_reader.cc:155]: Check!

**Comment by ahojnnes** [comment]
@Ambrosiussen There seems to be a bug in the logic, see: https://github.com/colmap/colmap/actions/runs/12295764100/job/34313358272?pr=3034

**Comment by Ambrosiussen** [comment]
> @Ambrosiussen There seems to be a bug in the logic, see: https://github.com/colmap/colmap/actions/runs/12295764100/job/34313358272?pr=3034

I think it actually exposed a bug! If you look at the dataset, and the arguments used for automatic reconstruction. No mask is supposed to be used at all!

Yet somehow it passes this check which was already there :)
`if (mask && !options_.mask_path.empty()) {`

`ubprocess.CalledProcessError: Command '['/home/runner/work/colmap/colmap/build/src/colmap/exe/colmap', 'automatic_reconstructor', '--image_path', 'boulders/images/', '--workspace_path', '/home/runner/work/colmap/colmap/eth3d_benchmark/boulders', '--use_gpu', '1', '--num_threads', '-1', '--quality', 'low', '--camera_model', 'PINHOLE', '--camera_params', '3408.29,3408.61,3115.6,2067.54']'`

**Comment by ahojnnes** [comment]
image_reader.cc:56 needs the following I would guess:
```
  if (options_.mask_path) {
    options_.mask_path =
        EnsureTrailingSlash(StringReplace(options_.mask_path, "\\", "/"));
  }
```

**Comment by Ambrosiussen** [comment]
@ahojnnes Any idea why these tests dont run?

**Comment by ahojnnes** [comment]
They need to be approved by a maintainer for first time contributors.

**Comment by ahojnnes** [comment]
Thank you for your contribution.

**Pull request merged**